instrumentation/httpurlconnection/testing/src/androidTest/kotlin/io/opentelemetry/instrumentation/library/httpurlconnection/InstrumentationTest.kt

```
@@ -55,12 +56,11 @@ class InstrumentationTest {
 
             executor.shutdown()
 
-            // Wait for all tasks to finish execution or timeout
-            assertThat(executor.awaitTermination(2, TimeUnit.SECONDS))
-                .withFailMessage("Test could not be completed as tasks did not complete within the 2s timeout period.")
-                .isTrue()
-
-            assertThat(openTelemetryRumRule.inMemorySpanExporter.finishedSpanItems.size).isEqualTo(4)
+            if (executor.awaitTermination(2, TimeUnit.SECONDS)) {
+                assertThat(openTelemetryRumRule.inMemorySpanExporter.finishedSpanItems.size).isEqualTo(4)
+            } else {
+                Log.e(TAG, "Test could not be completed as tasks did not complete within the 2s timeout period.")
```

in this case, the test should fail, or the test should be rewritten in a way that you can be notified once the tasks are completed

---

I'd use eg https://github.com/awaitility/awaitility/wiki/Usage

---

> The one I've changed it to for now - Prevent false negative by logging that execution did not complete within given max timeout, so you can go and increase timeout and run manually and test. Would adding a comment saying - test this manually and look for any logged error messages help?

its more about that this is run in CI and tests should break, no one is looking at logs on CI, so if its broken, CI is green and we will merge broken things.

> Increase the timeout (it is working in 4s now) and let it fail - this would mean the test is flaky (depends on the time the parallel execution of 4 requests take) and a false negative is possible.

increasing timeout is a workaround, and no solution, not telling we should not do this, but usually just incase testing time and the flakiness is still there, fixing the test (or even removing it) is usually better IMO

https://github.com/awaitility/awaitility/blob/master/awaitility-kotlin/src/test/kotlin/org/awaitility/kotlin/KotlinTest.kt
there are a few options, for example, `await().untilAsserted`, `await().until`, `await.forever until`, etc... so you don't specify a timer but rather a condition.

---

an example:

```kotlin
val executor = Executors.newFixedThreadPool(4)
        val counter = AtomicInteger(0)

        executor.submit {
            executeGet("http://httpbin.org/get")
            counter.incrementAndGet()
        }
        executor.submit {
            executeGet("http://google.com")
            counter.incrementAndGet()
        }
        executor.submit {
            executeGet("http://android.com")
            counter.incrementAndGet()
        }
        executor.submit {
            executeGet("http://httpbin.org/headers")
            counter.incrementAndGet()
        }

        executor.shutdown()

        await.untilAsserted {
            assertEquals(4, counter.get())
        }
```

---

another option is a `CountDownLatch(4)` calling `countDown()` for each of the submits.
and `cdl.await()` after

---

Infinite timeout in my opinion would be dangerous, it can be achieved with all 3 APIs - executor.await, latch.await, or await.forever, await.untilAsserted. But I did not want to go with it as I don't think we want to stall the test pipeline if in case anything were to go wrong with executing the test. 

By suggesting to log, I wanted to at least keep the code there for manual testing during our development (Mentioning that in function comments properly). I still think we can do this. 

Alternatively, we can remove the test if that's what everyone thinks is better? @LikeTheSalad @breedx-splk

---

I dont think its a problem since the executor is created in the test, so we have control over its lifecycle.
as far as I can tell, `executeGet` swallows all exceptions, so its guaranteed that eg `counter.incrementAndGet()` will be called.
the only difference is that this way you allow `executeGet` to take as much time as needed so its not flaky, if `executeGet` takes forever or does not finish for some reason, its an even more serious bug.

---

I'm with @marandaneto on this one, and I like the suggestion to use a `Latch` to accomplish fast success with a longer timeout to prevent flapping. There's more than one way to do this, but I put in a PR here to demonstrate: https://github.com/surbhiia/opentelemetry-android/pull/2

---

Thanks @marandaneto and @breedx-splk for your feedback and for taking the time to demonstrate the code changes. 

I still think it would be the same as `executor.awaitTermination(1, TimeUnit.MINUTES)` as `awaitTermination `is primarily concerned with waiting for task completion and doesn’t wait for the shutdown of the executor itself. 

As everyone prefers the `Latch`, I’ve updated the code as per the suggested changes now. :)